exercises/luhn/src/Example.hx

```diff
@@ -0,0 +1,33 @@
+package;
+
+using StringTools;
+using Lambda;
+
+class Luhn {
+    public static function valid(input: String): Bool {
+        // spaces are allowed but stripped
+        input = input.replace(" ", "");
+
+        // single digit is not valid
+        if (input.length <= 1)
+            return false;
+
+        // only digits are allowed
+        if (~/[^0-9]/.match(input))
+            return false;
+
+        // double every 2nd digit, subtracting 9 if result > 9
+        var digits = input.split("").map(Std.parseInt);
+        for (i in 0...Std.int(digits.length / 2)) {
+            var j = digits.length - (2 * i) - 2;
+            digits[j] = (digits[j] * 2);
+            if (digits[j] > 9)
+                digits[j] -= 9;
+        }
+
+        function sum(a, b) return a + b;
+        var digitSum = digits.fold(sum, 0);
+        
+        return digitSum % 10 == 0;
```

Comment:
I think this should be simplified to the following:

```haxe
        var sum = 0; 
        var second = false; //for tracking every other digit

        //loop in reverse order
        for (i in (1 - input.length)...1) {
            //charcodes of 0-9 are 48-57 
            var temp = input.charCodeAt(-i) - '0'.code; //convert char codes to digit value
            if (second) //double every other number 
                temp *= 2;

            //if temp is greater than 9 subtract 9
            sum += temp > 9 ? temp - 9 : temp;
            second = !second;
        }

        return sum % 10 == 0;
```

I ran it against the tests and they all pass.

Reply:
Agree, that looks easier to follow and clearer about its intent. Was a bit surprised to see it work using a backward/reverse iterator though, since natively only forward iteration is currently supported afaik (outside of a [custom iterator](https://code.haxe.org/category/data-structures/reverse-iterator.html)). I did a quick test and sure enough the compiler complained: 
```haxe
for (i in 5...0) trace(i); // Error: Cannot iterate backwards
```
But why is the code above working fine, I guess it's because the lower bound is not statically known in this case so at runtime it's silently turned into a no-op instead (might be for future-proofing):
```haxe
var x = 5;
for (i in x...0) trace(i); // [No output]
```
So the code ends up passing the tests because it's only testing `return sum % 10 == 0`m where sum still has its initial value of zero xd

Anyway, I re-worked the code to get around the lack of reverse iterators and learned something new. Thanks for the feedback!

Reply:
When iterating backwards you go off of negative numbers and then you reference negative i, `-i`.

From my code above:
```
for (i in (1 - input.length)...1) 

input.charCodeAt(-i)
```

Think of it this way: go forward with positives and go backwards with negatives.

### Foward/left to right (positive):
```
for(i in 0 ... array.length)
```

### Backwards/right to left (negative):
```
for (i in (1 - array.length) ... 1)
```

or

```
for (i in -(array.length - 1) ... 1)
```

Reply:
Aha! That's pretty clever and I completely missed it. Makes sense now :-)